lib/rules/func-names.js

```diff
@@ -147,24 +155,15 @@ module.exports = {
                 }
 
                 const hasName = Boolean(node.id && node.id.name);
+                const config = getConfigForNode(node);
 
-                if (node.generator) {
-                    const failsAsNeededRequirement = asNeededGenerators ? !hasInferredName(node) : !isObjectOrClassMethod(node);
-
-                    if (requireNamedGenerators && !hasName && failsAsNeededRequirement) {
-                        reportUnexpectedUnnamedFunction(node);
-                    } else if (!requireNamedGenerators && hasName) {
+                if (config === "never") {
+                    if (hasName) {
                         reportUnexpectedNamedFunction(node);
                     }
                 } else {
-                    if (never) {
-                        if (hasName) {
-                            reportUnexpectedNamedFunction(node);
-                        }
-                    } else {
-                        if (!hasName && (asNeeded ? !hasInferredName(node) : !isObjectOrClassMethod(node))) {
-                            reportUnexpectedUnnamedFunction(node);
-                        }
+                    if (!hasName && (config === "as-needed" ? !hasInferredName(node) : !isObjectOrClassMethod(node))) {
```

Comment:
Could this be simplified to something like:

```js
} else if (config === "as-needed") {
    if (!hasName && !hasInferredName(node)) {
        reportUnexpectedUnnamedFunction(node);
    }
} else {
    if (!isObjectOrClassMethod(node)) {
        reportUnexpectedUnnamedFunction(node);
    }
}
```

In my opinion, the expression `!hasName && (config === "as-needed" ? !hasInferredName(node) : !isObjectOrClassMethod(node))` is fairly difficult to read and understand (most of this complexity was already there before this change).